### PR TITLE
[DEV-4921] metric entries calculator batch calc jira grouping

### PR DESCRIPTION
**Pull request by gaetano-guerriero**

### CODE REVIEW

File: server/athenian/api/internal/features/metric_calculator.py [hunk 1]
```diff
@@ -1007,6 +1008,67 @@ def group_by_lines(lines: Sequence[int], column: np.ndarray) -> list[np.ndarray]
     return full_line_groups
 
 
+@dataclasses.dataclass(frozen=True, slots=True)
+class JIRAGroup:
```
Comment (vmarkovtsev) [inline server/athenian/api/internal/features/metric_calculator.py:1012]:
```suggestion
class JIRAGrouping:
```

File: server/athenian/api/internal/features/metric_calculator.py [hunk 2]
```diff
@@ -1007,6 +1008,67 @@ def group_by_lines(lines: Sequence[int], column: np.ndarray) -> list[np.ndarray]
...
+    projects: Optional[Collection[str]]
+    priorities: Optional[Collection[str]]
+    types: Optional[Collection[str]]
+
+    def __bool__(self) -> bool:
+        """Get whether the group definition is empty.  An empty group does not do any grouping."""
+        return any(f is not None for f in (self.projects, self.priorities, self.types))
+
+    @classmethod
+    def total(cls) -> JIRAGroup:
```
Comment (vmarkovtsev) [inline server/athenian/api/internal/features/metric_calculator.py:1029]:
```suggestion
    def empty(cls) -> JIRAGroup:
```
(to use the same name as we already have for JIRAFilter or LabelFilter)

File: server/athenian/api/internal/features/metric_calculator.py [hunk 3]
```diff
@@ -1007,6 +1008,67 @@ def group_by_lines(lines: Sequence[int], column: np.ndarray) -> list[np.ndarray]
...
+        """Return a `JIRAGroup` selecting all data."""
+        return cls(None, None, None)
```
Comment (vmarkovtsev) [inline server/athenian/api/internal/features/metric_calculator.py:1031]: btw, can we do the same thing by adding default values in the dataclass?
```python
class JIRAGrouping:
    projects: Optional[...] = None
```
I don't have a preference, just found out it was possible in the last PR.

File: server/athenian/api/internal/features/metric_calculator.py [hunk 4]
```diff
@@ -1007,6 +1008,67 @@ def group_by_lines(lines: Sequence[int], column: np.ndarray) -> list[np.ndarray]
...
+                    filter_values = np.array(list(group_props), dtype="S")
+                    all_values = np.concatenate(df_values)
+                    splits = np.cumsum(np.fromiter((len(v) for v in df_values), int, len(df)))
+                    match_mask = in1d_str(all_values, filter_values)
+                    match_indexes = np.flatnonzero(match_mask)
+                    field_res = np.unique(np.searchsorted(splits, match_indexes, side="right"))
+                    fields_res.append(field_res)
+
+            assert fields_res  # since bool(jira_group) is True
+            group_res = reduce(partial(np.intersect1d, assume_unique=True), fields_res)
```
Comment (vmarkovtsev) [inline server/athenian/api/internal/features/metric_calculator.py:1064]: Performance: we can avoid `intersect1d` and reduce memory if we do this:

```python
matched = np.zeros(len(df), dtype=np.uint8)
target = 0
for whatever in foo_bar:
    if group_props is not None:
        target += 1
        # ...
        matched[field_res] += 1
group_res = np.flatnonzero(matched == target)
```

File: server/athenian/api/internal/jira.py [hunk 5]
```diff
@@ -162,6 +173,60 @@ async def get_jira_installation_or_none(
...
+        self,
+        priority_names_map: dict[str, list[str]],
+        types_map: dict[str, list[str]],
+    ) -> None:
+        """Build a mapper using the given dictionaries as mapping source."""
+        self._priority_names_map = priority_names_map
+        self._types_map = types_map
+
+    @classmethod
+    async def build(cls, jira_acc_id: int, mdb: DatabaseLike) -> JIRAEntitiesMapper:
```
Comment (vmarkovtsev) [inline server/athenian/api/internal/jira.py:193]:
```suggestion
    async def load(cls, jira_acc_id: int, mdb: DatabaseLike) -> JIRAEntitiesMapper:
```
or `fetch()` - those two are used in the rest of the code.

File: server/athenian/api/internal/jira.py [hunk 6]
```diff
@@ -162,6 +173,60 @@ async def get_jira_installation_or_none(
...
+        for priority_name, priority_id in priority_rows:
+            priority_names_map[priority_name].append(priority_id)
+
+        type_stmt = (
+            sa.select(IssueType.name, IssueType.id)
+            .where(IssueType.acc_id == jira_acc_id)
+            .distinct()
+        )
+        types_map = defaultdict(list)
+        for type_, type_id in await mdb.fetch_all(type_stmt):
```
Comment (vmarkovtsev) [inline server/athenian/api/internal/jira.py:209]: Performance: must execute these two `mdb.fetch_all()` in parallel using `athenian.api.utils.gather()`

File: server/athenian/api/internal/jira.py [hunk 7]
```diff
@@ -162,6 +173,60 @@ async def get_jira_installation_or_none(
...
+        )
+        priority_rows = await mdb.fetch_all(priority_stmt)
+        priority_names_map = defaultdict(list)
+        for priority_name, priority_id in priority_rows:
+            priority_names_map[priority_name].append(priority_id)
+
+        type_stmt = (
+            sa.select(IssueType.name, IssueType.id)
+            .where(IssueType.acc_id == jira_acc_id)
+            .distinct()
```
Comment (vmarkovtsev) [inline server/athenian/api/internal/jira.py:206]: Do we really have duplicates there? :thinking:

**Review by vmarkovtsev — CHANGES_REQUESTED** [review]

File: server/athenian/api/internal/jira.py (hunk 7 above)
Comment (gaetano-guerriero) [reply server/athenian/api/internal/jira.py:206]: removed distinct from priority query
In IssueType we can, fixed our sqlalchemy schema

File: server/athenian/api/internal/jira.py [hunk 8]
```diff
@@ -162,6 +173,62 @@ async def get_jira_installation_or_none(
...
+    ) -> None:
+        """Build a mapper using the given dictionaries as mapping source."""
+        self._priority_names_map = priority_names_map
+        self._types_map = types_map
+
+    @classmethod
+    async def load(cls, jira_acc_id: int, mdb: DatabaseLike) -> JIRAEntitiesMapper:
+        """Load the mapper by loading data from DB."""
+        priority_stmt = sa.select(Priority.name, Priority.id).where(Priority.acc_id == jira_acc_id)
+        priority_rows = await mdb.fetch_all(priority_stmt)
```
Comment (vmarkovtsev) [inline server/athenian/api/internal/jira.py:196]: Forgot to remove this, we do the same in `gather()`

File: server/athenian/api/internal/features/metric_calculator.py [hunk 9]
```diff
@@ -1007,6 +1008,69 @@ def group_by_lines(lines: Sequence[int], column: np.ndarray) -> list[np.ndarray]
...
+    types: Optional[Collection[str]] = None
+
+    def __bool__(self) -> bool:
+        """Get whether the group definition is empty.  An empty group does not do any grouping."""
+        return any(f is not None for f in (self.projects, self.priorities, self.types))
+
+    @classmethod
+    def empty(cls) -> JIRAGrouping:
+        """Return a `JIRAGrouping` selecting all data."""
+        return cls(None, None, None)
```
Comment (vmarkovtsev) [inline server/athenian/api/internal/features/metric_calculator.py:1031]: No longer needed, we can call `JIRAGrouping()` directly.

File: server/athenian/api/internal/features/metric_calculator.py [hunk 10]
```diff
@@ -1007,6 +1008,69 @@ def group_by_lines(lines: Sequence[int], column: np.ndarray) -> list[np.ndarray]
     return full_line_groups
 
 
+@dataclasses.dataclass(frozen=True, slots=True)
+class JIRAGrouping:
+    """Grouping definition based on jira entities.
+
+    `None` field values indicate that no grouping should be done on the corresponding dataframe
+    property.
+    On the opposite an single empty field value makes this group select nothing.
```
Comment (vmarkovtsev) [inline server/athenian/api/internal/features/metric_calculator.py:1017]:
```suggestion
    On the opposite, a single empty field value makes this group select nothing.
```

**Review by vmarkovtsev — CHANGES_REQUESTED** [review]

File: server/athenian/api/internal/features/metric_calculator.py (hunk 9 above)
Comment (gaetano-guerriero) [reply server/athenian/api/internal/features/metric_calculator.py:1031]: https://github.com/athenianco/athenian-api/pull/2803#discussion_r972079702

File: server/athenian/api/internal/jira.py (hunk 8 above)
Comment (vmarkovtsev) [inline server/athenian/api/internal/jira.py:196]: Still not removed

**Review by vmarkovtsev — CHANGES_REQUESTED** [review]

**Review by vmarkovtsev — APPROVED** [review]

**Pull request merged**